device/trezor: HF10 protocol upgrade

ph4r05 opened the pull request:
HF10 protocol upgrade for Trezor.
Related to https://github.com/trezor/trezor-core/pull/478

I am sorry for a larger PR, I thought it is easier for you to review all at once as it is easier to see dependencies and why particular choices have been made.

## HF 10
- Most of the changes relate to the deterministic output masks in HF10. Masks are generated when a destination is sent to the Trezor, this affects how BP offloading is done. 
- Transaction Inputs Commitments and outputs commitments balancing is now performed in the signing step of the protocol, thus `pseudo_out`s are updated in the sign step as Trezor adjusts masks so their sums equal. 
- We discontinued support for Borromean range signatures so I removed the related logic from the client protocol code (offloading, verification, serialization).
- `in_memory` variant of the protocol which enabled to do signature a bit faster for standard transactions was removed. Trezor and Monero client code is now cleaner, safer (fewer branches). Speedup was not significant. 
- When all users upgrade to new Monero and Trezor firmwares, support for old protocol (`client_version=0`) can be dropped.
- Key image import after the cold signing imports only key images computed by the Trezor, i.e., key images of spent inputs.
- Minor refactoring of the Trezor client code (consts), passphrase entry refactored.
- key image import only for device-computed inputs

## Get TX key
- Trezor supports `get_tx_key` for decrypting stored tx private keys.
- `wallet2::get_tx_key_device()` calls `get_tx_key` on the Trezor if applicable
- Simplewallet supports `get_tx_key` and `get_tx_proof` on hw device using the `get_tx_key` feature

## Live refresh
- Live refresh feature enables to do the `refresh` with Trezor, i.e., computing key images on the fly. More convenient and efficient for users (`hw_key_image_sync` was required before).
- The first live refresh requires user confirmation on the device.  
- A thread is watching whether live refresh is being computed. If it is not for 30 seconds, it terminates the live refresh process - switches the Trezor state (display changes).

## Wallet API
- Initial Trezor support added so Monero GUI supports Trezor. 
- Wallet API Works with Monero-GUI (https://github.com/monero-project/monero-gui/pull/2019)
- Functional tests of the Wallet::API cold signing and refresh via mocked core RPC daemon (could be also used in different tests later). Mocker RPC daemon is created with the test `cryptonote::core` and runs on a local port. Wallet::API uses RPC daemon.

ph4r05 commented:
OK this PR is ready for review :)

ph4r05 commented:
@moneromooo-monero I've split code to more commits to make the review easier. I hope that helps.

Most of the code relates to the HF10 upgrade (and protocol simplification) + tests for HF9 and HF10.

I am wondering... as we have HF10/11 now, should I drop tests for HF9?

ph4r05 commented:
TODO:
- Adapt to #5277
- Adapt to #5267

binaryFate commented:
I find it a bit messy if some crypto functions like `hmac_keccak_init` are introduced with the sole purpose of Trezor integration, but are not identified as such by name or namespace, or at least by being in their own file.

Over time we may lose track of what crypto stuff has been introduced by who and for what purposes, and rely on it in a slightly different context without giving it the appropriate level of scrutiny.

binaryFate commented:
Also would it be possible to have files currently in  src/device_trezor/ to be part of a submodule? It is messy if we end up with 20 directories part of the main repo, each one for a different hardware provider.

Surely we can't move _everything_, but right now it seems like a lot of things ported to the Monero code base itself in a very intertwined manner.

ph4r05 commented:
> I find it a bit messy if some crypto functions like `hmac_keccak_init` are introduced with the sole purpose of Trezor integration, but are not identified as such by name or namespace, or at least by being in their own file.

I thought more like the HMAC is a universal and useful  crypto function which is easy to review if the underlying Keccak is correct. And it’s covered by tests. 

 Anyway, if you think that moving it out from the crypto to a separate folde then OK.
 Any suggestions on naming and names space ? @moneromooo-monero  ?

ph4r05 commented:
> Also would it be possible to have files currently in src/device_trezor/ to be part of a submodule? It is messy if we end up with 20 directories part of the main repo, each one for a different hardware provider.
> 
> Surely we can't move _everything_, but right now it seems like a lot of things ported to the Monero code base itself in a very intertwined manner.

 Moving to a submodule would bring some benefits, however  it could make the review a bit pain as the dependencies would be harder to spot.

 I would prefer to leave it in the main repository but if the other guys (e.g. @moneromooo-monero ) agree on the submodule variant and create the repo for it I will move it there.
 The question then is  how to review this PR.

moneromooo-monero commented on src/crypto/keccak.c:
```diff
@@ -212,3 +212,51 @@ void keccak_finish(KECCAK_CTX * ctx, uint8_t *md){
     }
 }
+
+void hmac_keccak_init(hmac_keccak_state *S, const uint8_t *_key, uint64_t keylen) {
```
Is this standard HMAC ? If so, it should probably in a hmac.[hc] file so it be used nicely by other stuff in the future.

ph4r05 replied:
Correct, it is standard HMAC RFC 2104. I've ported it from `src/crypto/blake256.c`checking also against the specification.

I will move it to another file and add comments it is RFC 2104 HMAC.

moneromooo-monero commented on src/crypto/keccak.c:
```diff
@@ -212,3 +212,51 @@ void keccak_finish(KECCAK_CTX * ctx, uint8_t *md){
+    memset(pad, 0x36, KECCAK_BLOCKLEN);
+    for (i = 0; i < keylen; ++i) {
+        pad[i] ^= key[i];
+    }
```
This allows keys with size < 32. That loop looks like a two byte key of 0x42 0x00 would behave the same as a one byte key of 0x42. Is that intended ?

ph4r05 replied:
I think this is according to the RFC 2104.

```
B = hash function block length
ipad = the byte 0x36 repeated B times
opad = the byte 0x5C repeated B times.
HMAC(K, text) = H(K XOR opad, H(K XOR ipad, text))
```

Also according to the section 3:
"The key for HMAC can be of any length (keys longer than B bytes are first hashed using H). 
 However, less than L (size of hash output) bytes is strongly discouraged as it would decrease the security strength of the function. "

moneromooo-monero commented on src/crypto/keccak.c:
```diff
@@ -212,3 +212,51 @@ void keccak_finish(KECCAK_CTX * ctx, uint8_t *md){
+    }
+    keccak_update(&S->outer, pad, KECCAK_BLOCKLEN);
+
+    memset(keyhash, 0, HASH_SIZE);
```
If this is intended to clear memory, please use memwipe instead.

moneromooo-monero commented on src/crypto/keccak.c:
```diff
@@ -212,3 +212,51 @@ void keccak_finish(KECCAK_CTX * ctx, uint8_t *md){
+    memset(keyhash, 0, HASH_SIZE);
+}
+
+void hmac_keccak_update(hmac_keccak_state *S, const uint8_t *data, uint64_t datalen) {
```
keccak_update takes a size_t. On 32 bit machines, you might plausibly manage an overflow to 0. Best to use size_t here too then I guess. Even tough you'd have to try to hash 4 GB, so unlikely.

moneromooo-monero commented on src/crypto/keccak.c:
```diff
@@ -212,3 +212,51 @@ void keccak_finish(KECCAK_CTX * ctx, uint8_t *md){
+    keccak_finish(&S->inner, ihash);
+    keccak_update(&S->outer, ihash, HASH_SIZE);
+    keccak_finish(&S->outer, digest);
+    memset(ihash, 0, 32);
```
memwipe

moneromooo-monero commented on src/crypto/keccak.c:
```diff
@@ -212,3 +212,51 @@ void keccak_finish(KECCAK_CTX * ctx, uint8_t *md){
+    memset(ihash, 0, 32);
+}
+
+void hmac_keccak_hash(uint8_t *out, const uint8_t *key, uint64_t keylen, const uint8_t *in, uint64_t inlen) {
```
I'd go for size_t here too.

moneromooo-monero commented:
MRL people are pro HMAC in general I believe, so if this is a real HMAC, it should be exposed in general. src/crypto is the right place for it, I'd put it in a separate file though.

About submodule or not for the trezor code, I think it's best to have it in tree since it's really closely intertwined. It's just a lot of code changes to review every time this changes for some reason ^_^

moneromooo-monero commented on src/device/device_cold.hpp:
```diff
@@ -44,13 +45,62 @@ namespace hw {
+      std::string salt2;
+      std::string tx_enc_keys;
+      std::string tx_prefix_hash;
+    } tx_key_data_t;
```
Some of those things look like secrets ? If so they should be crypto::secret_key or wipeable_string, which auto wipe on dtor.

ph4r05 replied:
These are just salts and ciphertexts, it is assumed public in this attacker model so it is maybe ok to be left as strings.

moneromooo-monero commented on src/device/device_cold.hpp:
```diff
@@ -65,6 +115,54 @@ namespace hw {
+
+    /**
+     * Decrypts TX keys.
+     * If view_public_key is set, derivations are computed instead of the tx private keys.
```
While they're the same size, key derivations would be better stored in a crypto::key_derivation, or it's doing to be massively confusing. Might be better to have two functions in this case (maybe calling the same internal code, I've not seen it yet).

ph4r05 replied:
True. I've removed this part and kept just tx_key computation as I noticed the intended use-case for derivations would require a bit change (tx_proof) which is not yet implemented in Trezor.

moneromooo-monero commented on src/device_trezor/device_trezor.cpp:
```diff
@@ -69,6 +70,67 @@ namespace trezor {
+      while(m_live_refresh_thread_running)
+      {
+        boost::this_thread::sleep_for(boost::chrono::milliseconds(10));
+        if (!m_transport || !m_live_refresh_in_progress)
```
That seems racy. There is no lock in release, which calls base::release, which calls disconnect, which deletes m_transport and sets it to nullptr.

ph4r05 replied:
Good catch, I've moved the release after thread join, so it waits for completion.

moneromooo-monero commented on src/device_trezor/device_trezor.cpp:
```diff
@@ -194,17 +289,22 @@ namespace trezor {
         MTRACE("Batch " << cur << " / " << num_batches << " batches processed");
+        EVENT_PROGRESS((double)cur * batch_size / mtds.size());
       }
+      EVENT_PROGRESS(1.);
```
Should you not call with 0 at function start ?

moneromooo-monero commented on src/device_trezor/device_trezor.cpp:
```diff
@@ -194,17 +289,22 @@ namespace trezor {
 
       for(auto & sub : kis){
         char buff[32*3];
+        size_t buff_len = 32*3;
```
sizeof(buff) ? Guards against later changes :)

moneromooo-monero commented on src/device_trezor/device_trezor.cpp:
```diff
@@ -194,17 +289,22 @@ namespace trezor {
                                           reinterpret_cast<const uint8_t *>(final_ack->enc_key().data()),
-                                          reinterpret_cast<const uint8_t *>(sub.iv().data()), buff);
+                                          reinterpret_cast<const uint8_t *>(sub.iv().data()), buff, &buff_len);
+        CHECK_AND_ASSERT_THROW_MES(buff_len == 32*3, "Plaintext size invalid");
```
Same

moneromooo-monero commented on src/device_trezor/device_trezor.cpp:
```diff
@@ -213,16 +313,144 @@ namespace trezor {
+        live_refresh_start();
+      }
+
+      m_last_live_refresh_time = std::chrono::system_clock::now();
```
You probably want to set this before calling live_refresh_start, or you might time out early if unlucky.

ph4r05 replied:
I've set it after the `live_refresh_start` because it sets `m_live_refresh_in_progress` which enables the thread auto-shutdown. The `live_refresh_start` is user-confirmed action so the idea was to set the `m_last_live_refresh_time` after this.

moneromooo-monero replied:
Nevermind, I confused m_live_refresh_running and m_live_refresh_in_progress.

moneromooo-monero commented on src/device_trezor/device_trezor.cpp:
```diff
@@ -213,16 +313,144 @@ namespace trezor {
+        if (!started && m_live_refresh_in_progress)
+        {
+          live_refresh_finish();
+        }
```
That's probably racy too, though I did not look at this one much :)

ph4r05 replied:
It could be OK, the `live_refresh_finish` has inside `AUTO_LOCK_CMD()`. If there is a race between shutdown thread one locks and finishes.

moneromooo-monero commented on src/device_trezor/device_trezor.cpp:
```diff
@@ -267,6 +495,10 @@ namespace trezor {
 
         // KI sync
+        for(size_t cidx=0, trans_max=unsigned_tx.transfers.second.size(); cidx < trans_max; ++cidx){
+          signed_tx.key_images[cidx] = unsigned_tx.transfers.second[cidx].m_key_image;
```
Is signed_tx.key_images guaranteed to have enough size at that place ?

ph4r05 replied:
Yes, it is set before for-loop
```cpp
signed_tx.key_images.resize(unsigned_tx.transfers.second.size());
```

moneromooo-monero commented on src/device_trezor/device_trezor.hpp:
```diff
@@ -69,11 +89,16 @@ namespace trezor {
       bool  has_tx_cold_sign() const override { return true; }
       void  set_network_type(cryptonote::network_type network_type) override { this->network_type = network_type; }
+      void  set_live_refresh_enabled(bool enabled) { m_live_refresh_enabled = enabled; }
+      bool  live_refresh_enabled() { return m_live_refresh_enabled; }
```
could be const

moneromooo-monero commented on src/device_trezor/device_trezor.cpp:
```diff
@@ -69,6 +70,67 @@ namespace trezor {
+        auto current_time = std::chrono::system_clock::now();
+        if (current_time - m_last_live_refresh_time.load() > std::chrono::seconds(20))
+        {
+          MTRACE("Closing live refresh process due to innactivity");
```
inactivity (sorry :P)

ph4r05 replied:
I've just fixed that 😄

moneromooo-monero commented on src/device_trezor/trezor/protocol.cpp:
```diff
@@ -33,13 +33,46 @@
+    throw std::invalid_argument("Field " #name " not found in JSON");\
+  } while(0)
+
+
```
That looks like it's copied from src/common. Can't you use it from there ?

ph4r05 replied:
I had to modify it a bit as the string can contain null-byte and this case was not covered by the src/common. I modified it a bit here so it creates a string with the length provided by the json lib

moneromooo-monero commented on src/device_trezor/trezor/protocol.cpp:
```diff
@@ -84,19 +117,23 @@ namespace protocol{
+
+    if (plaintext_len){
+      *plaintext_len = (size_t) res_len;
+    }
```
I suppose it could throw if it does not fit.

ph4r05 replied:
done

moneromooo-monero commented on src/device_trezor/trezor/protocol.cpp:
```diff
@@ -185,6 +222,49 @@ namespace ki {
+    CHECK_AND_ASSERT_THROW_MES(len_ciphertext > crypto::chacha::IV_SIZE + crypto::chacha::TAG_SIZE, "Invalid size");
+
+    size_t ki_len = len_ciphertext - crypto::chacha::IV_SIZE;
+    std::unique_ptr<uint8_t[]> plaintext(new uint8_t[ki_len]);
```
The plaintext will presumably not include the TAG_SIZE bytes ? Or maybe the assert above is technically wrong ?

ph4r05 replied:
fixed

moneromooo-monero replied:
Then there was another instance of this I did not point out since I was not sure I was not missing something since the same thing was done twice.

ph4r05 replied:
Correct, I've fixed that one also

moneromooo-monero commented on src/device_trezor/trezor/protocol.cpp:
```diff
@@ -267,9 +349,28 @@ namespace tx {
+
+    hmac_keccak_hash(hash, (const uint8_t *) salt.data(), salt.size(), hash, 32);
+    memcpy(res.data, hash, 32);
+    return res;
```
Looks like hash is used as a secret key, so you may want to memwipe it.

ph4r05 replied:
done

moneromooo-monero commented on src/device_trezor/trezor/protocol.cpp:
```diff
@@ -887,6 +1004,81 @@ namespace tx {
+
+    for(unsigned i = 0; i < keys_len / 32; ++i)
+    {
+      memcpy(tx_keys[i].data, plaintext.get() + 32 * i, 32);
```
plaintext should also be memwiped

ph4r05 replied:
done

moneromooo-monero commented on src/device_trezor/trezor/protocol.hpp:
```diff
@@ -92,11 +92,13 @@ namespace protocol{
 // Crypto / encryption
 namespace crypto {
 namespace chacha {
+  const unsigned IV_SIZE = 12;
```
That's going to be super confusing given the chacha iv size is 8 in chacha.h...

ph4r05 replied:
I will add a comment it is defined in RFC 7539. It defines 96-bit nonce/IV

moneromooo-monero commented on src/device_trezor/trezor/protocol.hpp:
```diff
@@ -215,6 +227,9 @@ namespace tx {
 
     void extract_payment_id();
     void compute_integrated_indices(TsxData * tsx_data);
+    bool should_compute_bp_now();
```
Can this be const ?

moneromooo-monero commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -6773,7 +6773,14 @@ bool simple_wallet::get_tx_key(const std::vector<std::string> &args_)
+  if (!found_tx_key)
+  {
+    found_tx_key = m_wallet->get_tx_key_device(txid, tx_key, additional_tx_keys);
+  }
```
Does this need to be one in the caller ?

ph4r05 replied:
I wanted to discuss this. How to do it correctly. 
`get_tx_key` is const so I cannot just fallback to device in the `wallet::get_tx_key` if it needs to be loaded from the device.

moneromooo-monero replied:
Sounds like the device fields should be mutable then. Good old C++ bodge ? :)

ph4r05 replied:
Soo do you prefer `wallet2::get_tx_key` is not `const` anymore and does the device call if needed? I would like it more. 

Or I can rename `get_tx_key` to `get_tx_key_stored` which would be const, the `get_tx_key`  would do `get_tx_key_stored` first and then query the device. Does that sound OK?

moneromooo-monero replied:
It does sound OK, thanks. Though "_stored" is not very enlightening as to what it does, but it'll do.

ph4r05 replied:
Yeah, I sometimes struggle with names, maybe cached? :)

moneromooo-monero commented on src/wallet/api/wallet.cpp:
```diff
@@ -242,6 +242,42 @@ struct Wallet2CallbackImpl : public tools::i_wallet2_callback
+      return boost::none;
+    }
+
+    virtual void on_device_progresst(const hw::device_progress & event)
```
progresst ?

ph4r05 replied:
Good catch!

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -9873,6 +9884,70 @@ bool wallet2::get_tx_key(const crypto::hash &txid, crypto::secret_key &tx_key, s
+    THROW_WALLET_EXCEPTION_IF(tx_hash != txid, error::wallet_internal_error,
+                              "Failed to get the right transaction from daemon");
+
+    tx_key_data.tx_prefix_hash = std::string(tx_prefix_hash.data, 32);
```
There is a get_pruned_tx function in that file which should do the above.

ph4r05 replied:
Hmm I am not sure how I could use it. I need `transaction_prefix_hash`, the `get_pruned_tx` is maybe a bit more complicated?

moneromooo-monero replied:
OK, no problem.

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -11448,29 +11529,48 @@ uint64_t wallet2::import_key_images(const std::vector<std::pair<crypto::key_imag
     td.m_key_image_request = false;
     td.m_key_image_partial = false;
-    m_pub_keys[m_transfers[i].get_public_key()] = i;
+    m_pub_keys[td.get_public_key()] = i;
```
That seems wrong. Those are supposed index into m_transfers. Are you sure you don't mean to use "= idx" on these two ?

ph4r05 replied:
The idea is that `key_images` can be shorter than `m_transfers`, specified by the offset, like with `export_outputs`.

```cpp
const size_t idx = i + offset;
transfer_details &td = m_transfers[idx];
```

moneromooo-monero replied:
Yes, but here:
m_pub_keys[td.get_public_key()] = i;

That i is going to offset m_transfers, it's supposed to point back to the transfer_details structure. AFAICT, it's going to point to the wrong one here.

ph4r05 replied:
oh, I missed that. You are right! Thanks!

ph4r05 replied:
I've changed var names to ki_idx and transfer_idx so it is clear now

moneromooo-monero commented on src/wallet/wallet2.cpp:
```diff
@@ -11448,29 +11529,48 @@ uint64_t wallet2::import_key_images(const std::vector<std::pair<crypto::key_imag
+bool wallet2::import_key_images(std::vector<crypto::key_image> key_images, size_t offset, boost::optional<std::unordered_set<size_t>> selected_transfers)
 {
-  if (key_images.size() > m_transfers.size())
+  if (key_images.size() + offset > m_transfers.size())
```
Is this offset coming from a trusted source (ie, the code) or the device ? It the latter, this should check for overflow.

ph4r05 replied:
The offset is from the wallet. It is like the offset that is generated by `wallet2::export_outputs`

ph4r05 commented:
btw if you want to discuss something on IRC pls ping me on monero-devs. My IRC relay is somehow broken and does not store direct messages.

moneromooo-monero commented on tests/unit_tests/hmac_keccak.cpp:
```diff
@@ -0,0 +1,173 @@
+    },
+    {(char*)"00112233445566778899aabbccddeeff102132435465768798a9bacbdcedfe0f",
+     (char*)"000d1a2734414e5b6875828f9ca9b6c3d0ddeaf704111e2b3845525f6c798693a0adbac7d4e1eefb0815222f3c495663707d8a97a4b1becbd8e5f2ff0c192633404d5a6774818e9ba8b5c2cfdce9f603101d2a3744515e6b7885929facb9c6d3e0edfa0714212e3b4855626f7c8996a3b0bdcad7e4f1fe0b1825323f4c596673808d9aa7b4c1cedbe8f5020f1c293643505d6a7784919eabb8c5d2dfecf90613202d3a4754616e7b8895a2afbcc9d6e3f0fd0a1724313e4b5865727f8c99a6b3c0 ...
+     (char*)"cbe3ca627c6432c9a66f07c6411fccca894c9b083b55d0773152460142a676ed",
```
Can't those all be const in the first place ?

moneromooo-monero commented on tests/unit_tests/hmac_keccak.cpp:
```diff
@@ -0,0 +1,173 @@
+    },
+};
+
+#define TEST_KECCAK(sz, chunks) \
```
This can be removed

ph4r05 commented on tests/trezor/trezor_tests.cpp:
```diff
@@ -483,8 +563,18 @@ static std::vector<tools::wallet2*> vct_wallets(tools::wallet2* w1=nullptr, tool
+static uint64_t get_available_funds(tools::wallet2* wallet, uint32_t account=0)
+{
+  uint64_t sum = 0;
+  for(const auto & cur : wallet_accessor_test::get_transfers(wallet))
```
Ah thats true, I dont need this particular accessor here. 
`wallet_accessor_test` returns non-const, I need it somewhere else for modifications (tests), but here the const is OK. I will change it

ph4r05 commented on tests/trezor/daemon.cpp:
```diff
@@ -0,0 +1,369 @@
+  tools::options::set_option(vm, cryptonote::core_rpc_server::arg_rpc_bind_port, po::variable_value(std::to_string(initial_port + 1), false));
+  tools::options::set_option(vm, daemon_args::arg_zmq_rpc_bind_port, po::variable_value(std::to_string(initial_port + 2), false));
+  po::notify(vm);
+}
```
The notify? No particular reason here, I've just noticed the vm change code usually calls the notification.

moneromooo-monero replied:
I mean, why random ports ?

ph4r05 replied:
I’ve experimented with multiple running daemons but it is probably safer to let user specify the number to avoid possible random clashes

ph4r05 commented on tests/trezor/daemon.h:
```diff
@@ -0,0 +1,157 @@
+  static void randomize_ports(boost::program_options::variables_map & vm);
+
+  mock_daemon * set_start_p2p(bool fl) { m_start_p2p = fl || m_start_zmq; return this; }
+  mock_daemon * set_start_zmq(bool fl) { m_start_zmq = fl; m_start_p2p = m_start_p2p || fl; return this; }
```
I mean, the desired functionality is enabled zmq => enabled p2p. Or would it be better to leave this as ordinary setters and check for consistency on start?

moneromooo-monero replied:
I think it'd certainly be less confusing, but if you prefer if that way I'm ok with it too.

ph4r05 replied:
done

ph4r05 commented:
So I've analyzed it with [Coverity](https://scan.coverity.com/projects/ph4r05-monero) and it seems fine. (If you cannot see the analysis result I can invite you to the project).

Is there anything left for me to do on this PR? :)

ph4r05 commented:
Hmm Iam trying clang sanitizers, the race in live_refresh monitor thread is still there according to the thread sanitizer. I am fixing it now...

moneromooo-monero commented:
I had finished the review. It can go in next.

ph4r05 commented:
I've did small fix on the thread and the race on `m_transport`

ph4r05 commented:
> I had finished the review. It can go in next.

Great, thanks a lot! So now just a formal PR approve?

ph4r05 commented:
Meh. During the testing, I've noticed a small race in live refresh monitor. I've simplified the locking logic and fixed mutex locking. Now tests are OK for me. That should be all now.

Thanks for the check!

ph4r05 commented:
@fluffypony thanks! 🎉 🍻
